pkg/distributor/distributor.go

```diff
@@ -1006,36 +1010,151 @@ func (d *Distributor) metricsMiddleware(next push.Func) push.Func {
 	}
 }
 
+type ctxKey int
+
+const requestStateKey ctxKey = 1
+
+// requestState represents state of checks for given request. If this object is stored in context,
+// it means that request has been checked against inflight requests limit, and FinishPushRequest,
+// cleanupAfterPushFinished (or both) must be called for given context.
+type requestState struct {
+	// If set to true, push request will perform cleanup of inflight metrics after request has actually finished
+	// (which can be after push handler returns).
+	pushHandlerPerformsCleanup bool
+
+	// If positive, it means that request size has been checked and added to inflightPushRequestsBytes.
+	requestSize int64
+}
+
+// StartPushRequest does limits checks at the beginning of Push request in distributor.
+//
+// This can be called from different places, even multiple times for the same request:
+//
+// * from gRPC Method limit check. This only applies if request arrived as httpgrpc.HTTPRequest, and request metadata in gRPC request provided enough information to do the check.
+//
+// * from Distributor's limitsMiddleware method.
+//
+// This method creates requestState object and stores it in the context. This object describes which checks were already performed on the request,
+// and which component is responsible for doing a cleanup.
+//
+// This method returns errors with gRPC status code.
+func (d *Distributor) StartPushRequest(ctx context.Context, requestSize int64) (context.Context, error) {
+	ctx, _, err := d.startPushRequest(ctx, requestSize)
+	if err != nil {
+		return ctx, status.Error(codes.Unavailable, err.Error())
+	}
+	return ctx, err
+}
+
+func (d *Distributor) startPushRequest(ctx context.Context, requestSize int64) (context.Context, *requestState, error) {
+	// If requestState is already in context, it means that StartPushRequest already ran for this request.
+	rs, alreadyInContext := ctx.Value(requestStateKey).(*requestState)
+	if alreadyInContext {
+		return ctx, rs, nil
+	}
+
+	cleanupInDefer := true
+
+	// Increment number of requests and bytes before doing the checks, so that we hit error if this request crosses the limits.
+	inflight := d.inflightPushRequests.Inc()
+	defer func() {
+		if cleanupInDefer {
+			d.inflightPushRequests.Dec()
```

Comment:
I'm wondering if to keep the code more future proof, we should decrease `d.inflightPushRequestsBytes` here. My take is that in in the future we'll add more checks after `d.checkRequestSize()` then we'll miss subtracting it. The check here could be done on `rs.requestSize`.

Reply:
Actually we could just change the defer function into a call to `cleanupAfterPushFinished()`.